src/cache/cache.rs

```diff
@@ -0,0 +1,186 @@
+    CacheCapacityIsTooSmall,
+    CacheFull,
+}
+
+#[derive(Debug, Clone)]
+pub struct BytesCacheEntry {
+    pub bytes: Bytes,
+}
+
+// Individual entry destined for the byte cache.
+impl BytesCacheEntry {
+    pub fn new(bytes: Bytes) -> Self {
+        Self { bytes }
+    }
+}
+
+pub struct BytesCache {
+    bytes_table: IndexMap<String, BytesCacheEntry>,
+    total_capacity: usize,
+    entry_size_limit: usize,
+}
+
+impl BytesCache {
+    pub fn new() -> Self {
+        Self {
+            bytes_table: IndexMap::new(),
+            total_capacity: 0,
+            entry_size_limit: 0,
+        }
+    }
+
+    // With a total capacity in bytes.
+    pub fn with_capacity(cap: usize) -> Self {
+        let mut new = Self::new();
+
+        new.total_capacity = cap;
+
+        new
+    }
+
+    // With a limit for individual entry sizes.
+    pub fn with_entry_size_limit(esl: usize) -> Self {
+        let mut new = Self::new();
+
+        new.entry_size_limit = esl;
+
+        new
+    }
+
+    // With bot a total capacity limit and an individual entry size limit.
+    pub fn with_capacity_and_entry_size_limit(cap: usize, esl: usize) -> Result<Self, Error> {
+        if esl > cap {
+            return Err(Error::EntrySizeLimitExceedsTotalCapacity);
+        }
+
+        let mut new = Self::new();
+
+        new.total_capacity = cap;
+        new.entry_size_limit = esl;
+
+        Ok(new)
+    }
+
+    pub async fn get(&self, key: &str) -> Option<BytesCacheEntry> {
+        self.bytes_table.get(key).cloned()
+    }
+
+    // Return the amount of entries in the map.
+    pub async fn len(&self) -> usize {
+        self.bytes_table.len()
+    }
+
+    // Size of all the entry bytes combined.
+    pub fn total_size(&self) -> usize {
+        let mut size: usize = 0;
+
+        for (_, entry) in self.bytes_table.iter() {
+            size += entry.bytes.len();
+        }
+
+        size
+    }
+
+    // Insert bytes using key.
+    // TODO: Freed space might need to be reserved. Hold and pass write lock between functions?
+    // For TO DO above: semaphore: Arc<tokio::sync::Semaphore>, might be a solution.
+    pub async fn set(&mut self, key: String, bytes: Bytes) -> Result<Option<BytesCacheEntry>, Error> {
+        if bytes.len() > self.entry_size_limit {
+            return Err(Error::BytesExceedEntrySizeLimit);
+        }
+
+        // Remove the old entry so that a new entry will be added as last in the queue.
+        let _ = self.bytes_table.shift_remove(&key);
+
+        let bytes_cache_entry = BytesCacheEntry::new(bytes);
+
+        self.free_size(bytes_cache_entry.bytes.len())?;
+
+        Ok(self.bytes_table.insert(key, bytes_cache_entry))
+    }
+
+    // Free space. Size amount in bytes.
+    fn free_size(&mut self, size: usize) -> Result<(), Error> {
+        // Size may not exceed the total capacity of the bytes cache.
+        if size > self.total_capacity {
+            return Err(Error::CacheCapacityIsTooSmall);
```

Comment:
Hey @josecelano ,

I agree that `CannotFreeMoreSpaceThanCapacity` would fit better in this case.

I think that this should be an error, cause when you instruct the cache to free a certain amount of bytes and it can't do it for whatever reason. The calling function should know why. In this case because the total cache size is smaller than the amount of bytes you want freed.